### DISCUSSION TITLE
6.3.0 Bad Gateway on cf_auth

### DISCUSSION THREAD

**Discussion by gaigaslab-operations** [concourse/concourse#5832] [Help & Support]
## Summary

Upgraded from 6.0.0 to 6.3.0
cf_auth fails at last redirect step with 502 - Bad Gateway

Starting from redirect back from CF login.<sys-domain>

/sky/issuer/callback?code=<redacted> (303)
/sky/issuer/approval?req=<redacted> (303)
/sky/callback?code (502) **Should be a (307) based on 6.0.0 behavior**

## Steps to reproduce

- Setup a team with cf_auth
- Login
- After authentication to cf, 502 error

## Expected results
To be logged in

## Actual results
502 Bad Gateway error page

## Triaging info

* Concourse version:
* Browser (if applicable):
* Did this used to work?

**Comment by jamieklassen** [comment]
Converting this to a discussion until we can find a consistent reproducible, following https://blog.concourse-ci.org/nobody-wants-a-stale-bot/#an-update-on-triage. I'm going to try to reproduce this. My first attempt, using https://github.com/cloudfoundry-incubator/cfdev, failed pretty hard probably because of my relatively-underpowered laptop. Next up I will try to get access to a shared cloud foundry instance managed by somebody else.

**Comment by gaigaslab-operations** [comment]
I am free to do a conference call next week as this is high priority before
we deploy to our developers application pipeline instance.

**Comment by xtreme-sameer-vohra** [comment]
Hi @gaigaslab-operations. @aoldershaw and I tried to reproduce this today with a locally deployed (`docker-compose`d) Concourse (`concourse/concourse:6.3.0-ubuntu`) but were unable to.

We started by creating a client with UAA:

```
$ uaac client add --name localconcourse --scope openid,cloud_controller.read \
  --authorized_grant_types "authorization_code,refresh_token" \
  --access_token_validity 3600 --refresh_token_validity 3600 \
  --redirect_uri http://localhost:8080/sky/issuer/callback
Client ID:  localconcourseclientid
  scope: cloud_controller.read openid
  client_id: localconcourseclientid
  resource_ids: none
  authorized_grant_types: refresh_token authorization_code
  redirect_uri: http://localhost:8080/sky/issuer/callback
  autoapprove:
  access_token_validity: 3600
  refresh_token_validity: 3600
  authorities: uaa.none
  name: localconcourse
  required_user_groups:
  lastmodified: 1594060947000
  id: localconcourseclientid
```

We then used the following `docker-compose` override to configure CF auth:

```yaml
version: '3'
services:
  web:
    environment:
      CONCOURSE_CF_API_URL: https://api.my-cf-deployment.com
      CONCOURSE_CF_CLIENT_ID: localconcourseclientid
      CONCOURSE_CF_CLIENT_SECRET: localconcourseclientsecret
      CONCOURSE_CF_SKIP_SSL_VALIDATION: "true"
```

We created a user in CF and assigned them to an org and space:

```
$ cf create-user concourse-user password
$ cf create-space concourse
$ cf set-org-role concourse-user system OrgManager
$ cf set-space-role concourse-user system concourse SpaceManager
```

And set several teams in concourse:

```
$ fly -t dev set-team -n org --cf-org system
$ fly -t dev set-team -n space --cf-space system:concourse
$ fly -t dev set-team -n user --cf-user concourse-user
```

Then, logging in to Concourse via CF auth with the configured credentials works (notably, the final `/sky/callback?code` returns a 307)

Is there anything in the Concourse (or more specifically, skymarshal) logs about what is actually failing on `/sky/callback`? And does the 502 response have a response body?

Also, what version of CloudFoundry are you using?

**Comment by xtreme-sameer-vohra** [comment]
Also, just clarifying, were you using a client that you created similar to the one above ?

**Comment by gaigaslab-operations** [comment]
I followed the same steps:

My UAA Client (note that I have dashes and dots in my cf client_id
(cf_auth.client_id).

> uaac client get concourse-platform-1.gaig.com
>   scope: cloud_controller.read openid
>   client_id: concourse-platform-1.gaig.com
>   resource_ids: none
>   authorized_grant_types: refresh_token authorization_code
>   redirect_uri: https://concourse-platform-1.gaig.com/sky/issuer/callback
>   autoapprove: true
>   access_token_validity: 86400
>   refresh_token_validity: 86400
>   authorities: uaa.none
>   name: concourse-platform-1.gaig.com
>   lastmodified: 1551826732000
>

My BOSH config:

>       cf_auth:
>         client_id: ((client_id))
>         client_secret: ((client_secret))
>         api_url: https://api.sys.cfnp-1.gaig.com

        skip_ssl_validation: true


My team setup:

flynew -t c4p gt -n platform
> name/role       users            groups
> platform/owner  local:<redacted>  cf:utility:platform:developer




502 Callback har snippet attached.

<html><body><h1>502 Bad Gateway</h1>
The server returned an invalid or incomplete response.
</body></html>

**Reply by aoldershaw** [reply]
The main difference I'm seeing here is `autoapprove: true`, and am wondering if that's causing issues perhaps? Not a UAA expert so not sure exactly what the implications of `autoapprove` are

**Comment by gaigaslab-operations** [comment]
Where would I find the skymarshall logs?

**Comment by gaigaslab-operations** [comment]
I see...in a BOSH deployment, all logs are combined.

I see where it enumerated my groups (too many to post here), but
`utility:platform:developer` is in there.

Here's what I think is the failure log:
{"timestamp":"2020-07-06T20:52:13.142810807Z","level":"error","source":"atc","message":"atc.sky.callback.failed-to-fetch-dex-token","data":{"error":"oauth2:
cannot fetch token: 400 Bad Request\nResponse:
{\"error\":\"invalid_request\",\"error_description\":\"Invalid or expired
code parameter.\"}","session":"16.93"}}

**Reply by aoldershaw** [reply]
Looking at `dex`, this error means that the auth code was not found in the storage backend (I believe that's Postgres in our case):

https://github.com/concourse/dex/blob/dcbe94f28c882227900fe6cfdc65acdf505f19c6/server/handlers.go#L715-L724

```go
	authCode, err := s.storage.GetAuthCode(code)
	if err != nil || s.now().After(authCode.Expiry) || authCode.ClientID != client.ID {
		if err != storage.ErrNotFound {
			s.logger.Errorf("failed to get auth code: %v", err)
			s.tokenErrHelper(w, errServerError, "", http.StatusInternalServerError)
		} else {
			s.tokenErrHelper(w, errInvalidRequest, "Invalid or expired code parameter.", http.StatusBadRequest)
		}
		return
	}
```

Will have to dig into the dex code a bit more to figure out why that could be the case

**Comment by gaigaslab-operations** [comment]
That's for scopes as in (resource).(scope).

**Comment by gaigaslab-operations** [comment]
Some kind of a race condition?

**Comment by xtreme-sameer-vohra** [comment]
Hey @gaigaslab-operations,
On your end are you able to validate
- whether the issue occurs on a new stood up 6.3.0 environment ? 
- does the issue occur for a new user in CF and new team in Concourse ?
We'd like to narrow down the cause to understand whether it is related to an upgrade or potentially due to the configuration.

I tried to repro the steps by performing an upgrade 6.0.0 -> 6.3.0 using a docker setup with a single team, which has a single user and I wasn't able to reproduce the issue. 

Steps on concourse repo
1.  git checkout v6.0.0
1. Edit docker-compose file to add 
    ```
      CONCOURSE_CF_API_URL: https://api.sys.SOME_CF_DOMAIN.com
      CONCOURSE_CF_CLIENT_ID: SOME_CLIENT_ID
      CONCOURSE_CF_CLIENT_SECRET: "SOME_CLIENT_SECRET
      CONCOURSE_CF_SKIP_SSL_VALIDATION: "true"
    ```
1. docker-compose up --build -d
1. fly -t LOCAL_ENV set-team -n some-team --cf-user=concourse-user
    ```
    name/role     users              groups
    some-team/owner  cf:concourse-user  none
    ```
1. Login via browser using `concourse-user`
1. git checkout v6.3.0
1. docker-compose up --build -d
1. Login via browser using `concourse-user`

**Reply by gaigaslab-operations** [reply]
Hello, 

We are not setup for docker here.

I performed a fresh BOSH install of 6.3.0.  This is using a different uaa client - concourse-test.gaig.com

BOSH Deployment Info:
```
c4test               bpm/1.1.8                      bosh-vsphere-esxi-ubuntu-xenial-go_agent/621.74  -  
                     concourse/6.3.0                                                                   
                     haproxy/9.3.0                                                                     
                     postgres/41                                                                       
                     ulimit/1                                                                          

```
I created a team:
flynew -t c4test set-team -n platform --local-user=cpipeuser --cf-space=utility:platform

I logged in with the internal user - ALL GOOD
I logged in with CF Authentication - 502 Bad Gateway

UAA Version Info on our PCF/PAS/TAS (they keep changing the name)
uaa/74.5.17

Remember that we still have a 6.0.0 environment that is using the same CF Foundation and same uaa client configuration (different client, same configuration).

6.0.0
```
uaac client get concourse-app-1.gaig.com
  scope: cloud_controller.read openid
  client_id: concourse-app-1.gaig.com
  resource_ids: none
  authorized_grant_types: refresh_token authorization_code
  redirect_uri: https://concourse-app-1.gaig.com/sky/issuer/callback
  autoapprove: true
  access_token_validity: 86400
  refresh_token_validity: 86400
  authorities: cloud_controller.admin
  name: concourse-app-1.gaig.com
  lastmodified: 1553384771000
```

6.3.0 client
```
uaac client get concourse-test.gaig.com

  scope: cloud_controller.read openid
  client_id: concourse-test.gaig.com
  resource_ids: none
  authorized_grant_types: refresh_token authorization_code
  redirect_uri: https://concourse-test.gaig.com/sky/issuer/callback
  autoapprove: true
  access_token_validity: 86400
  refresh_token_validity: 86400
  authorities: cloud_controller.admin
  name: concourse-test.gaig.com
  lastmodified: 1549650913000
```

I'm open for suggestions.

**Comment by gaigaslab-operations** [comment]
I'll try.

**Comment by gaigaslab-operations** [comment]
Ok, some more information.

I deleted the c4test 6.3.0 fresh deployment and reverted with everything the same to 6.0.0.  
That worked perfectly! This verified that my deployment, uaa client, and all other components were operational and represented the pre-upgrade configuration.  I could successfully login with a local user and a cf user in the same utility:platform org:space as a SpaceDeveloper.

Then I deleted the 6.0.0 deployment.

Then I deployed a fresh 6.1.0 deployment.  This broke with the same 502 Bad Gateway, so it looks like the problem was introduced in 6.1.0.

Then I deleted the 6.1.0 deployment, deployed 6.0.0 and upgraded to 6.1.0.  This worked on 6.0.0 again, but broke after upgrading to 6.1.0.

I also tried a fresh deployment of 6.2.0 and received the same 502 Bad Gateway.
I also tried a 6.0.0 upgraded to 6.2.0 and received the same 502 Bad Gateway.

So, it appears that everything after 6.0.0 reproduces the 502 Bad Gateway issue with cf_auth.

Hopefully this helps.

**Reply by aoldershaw** [reply]
Hey @gaigaslab-operations,

We're struggling to reproduce this issue on our end. One notable change with 6.1.0 is that the size of tokens is now proportional to the number of teams a user a member of. We've seen several issues stemming from this increased token size.

Could you try creating a new CF user that isn't a member of any spaces and logging in with that user?

How is network ingress controlled to the ATC? Is there any load balancer configuration that could potentially e.g. truncate cookies?

**Reply by gaigaslab-operations** [reply]
Inbound is through a BOSH-deployed HAProxy instance

**Reply by aoldershaw** [reply]
As a workaround before we fix https://github.com/concourse/concourse/issues/5851, you could maybe try setting `LimitRequestFieldSize` on your HAProxy to something larger than the default of `8192`. Unconfirmed that this will fix the problem, though.

**Reply by gaigaslab-operations** [reply]
@aoldershaw 
I used the directive ` buffer_size_bytes: 65536` in our bosh deployment for haproxy to workaround this issue.  I am now able to login with my cf credentials.

I'll be watching for fix to #5851,

**Reply by aoldershaw** [reply]
@gaigaslab-operations thanks, that's good to know about the `buffer_size_bytes` directive - I'm glad you're at least able to workaround the issue for the time being

**Comment by gaigaslab-operations** [comment]
I was able to login with a new user assigned to a single org/space.  It seems to be related to the DEX message sizing issue.
